Libc/gen/rand48.3
.\" Copyright (c) 1993 Martin Birgmeier
.\" All rights reserved.
.\"
.\" You may redistribute unmodified or modified versions of this source
.\" code provided that the above copyright notice and this and the
.\" following conditions are retained.
.\"
.\" This software is provided ``as is'', and comes with no warranties
.\" of any kind. I shall in no event be liable for anything that happens
.\" to anyone/anything when using this software.
.\"
.\"     @(#)rand48.3 V1.0 MB 8 Oct 1993
.\" $FreeBSD: src/lib/libc/gen/rand48.3,v 1.16 2004/07/02 23:52:10 ru Exp $
.\"
.Dd October 8, 1993
.Dt RAND48 3
.Os
.Sh NAME
.Nm drand48 ,
.Nm erand48 ,
.Nm jrand48 ,
.Nm lcong48 ,
.Nm lrand48 ,
.Nm mrand48 ,
.Nm nrand48 ,
.Nm seed48 ,
.Nm srand48
.Nd pseudo random number generators and initialization routines
.Sh LIBRARY
.Lb libc
.Sh SYNOPSIS
.In stdlib.h
.Ft double
.Fo drand48
.Fa void
.Fc
.Ft double
.Fo erand48
.Fa "unsigned short xsubi[3]"
.Fc
.Ft long
.Fo jrand48
.Fa "unsigned short xsubi[3]"
.Fc
.Ft void
.Fo lcong48
.Fa "unsigned short param[7]"
.Fc
.Ft long
.Fo lrand48
.Fa void
.Fc
.Ft long
.Fo mrand48
.Fa void
.Fc
.Ft long
.Fo nrand48
.Fa "unsigned short xsubi[3]"
.Fc
.Ft "unsigned short *"
.Fo seed48
.Fa "unsigned short seed16v[3]"
.Fc
.Ft void
.Fo srand48
.Fa "long seedval"
.Fc
.Sh DESCRIPTION
The
.Fn rand48
family of functions generates pseudo-random numbers, using a linear
congruential algorithm working on integers 48 bits in size.
The
particular formula employed is
r(n+1) = (a * r(n) + c) mod m.
The default value for the multiplicand `a' is 0xfdeece66d (25214903917).
The default value for the the addend `c' is 0xb (11).
The modulo is always fixed at m = 2 ** 48.
r(n) is called the seed of the random number generator.
.Pp
For the six generator routines described next, the first
computational step is to perform a single iteration of the algorithm.
.Pp
The
.Fn drand48
and
.Fn erand48
functions
return values of type double.
The full 48 bits of r(n+1) are
loaded into the mantissa of the returned value, with the exponent set
such that the values produced lie in the interval [0.0, 1.0).
.Pp
The
.Fn lrand48
and
.Fn nrand48
functions
return values of type long in the range
[0, 2**31-1].
The high-order (31) bits of
r(n+1) are loaded into the lower bits of the returned value, with
the topmost (sign) bit set to zero.
.Pp
The
.Fn mrand48
and
.Fn jrand48
functions
return values of type long in the range
[-2**31, 2**31-1].
The high-order (32) bits of
r(n+1) are loaded into the returned value.
.Pp
The
.Fn drand48 ,
.Fn lrand48 ,
and
.Fn mrand48
functions
use an internal buffer to store r(n).
For these functions
the initial value of r(0) = 0x1234abcd330e = 20017429951246.
.Pp
On the other hand,
.Fn erand48 ,
.Fn nrand48 ,
and
.Fn jrand48
use a user-supplied buffer to store the seed r(n),
which consists of an array of 3 shorts, where the zeroth member
holds the least significant bits.
.Pp
All functions share the same multiplicand and addend.
.Pp
The
.Fn srand48
function
is used to initialize the internal buffer r(n) of
.Fn drand48 ,
.Fn lrand48 ,
and
.Fn mrand48 ,
such that the 32 bits of the seed value are copied into the upper 32 bits
of r(n), with the lower 16 bits of r(n) arbitrarily being set to 0x330e.
Additionally, the constant multiplicand and addend of the algorithm are
reset to the default values given above.
.Pp
The
.Fn seed48
function
also initializes the internal buffer r(n) of
.Fn drand48 ,
.Fn lrand48 ,
and
.Fn mrand48 ,
but here all 48 bits of the seed can be specified in an array of 3 shorts,
where the zeroth member specifies the lowest bits.
Again,
the constant multiplicand and addend of the algorithm are
reset to the default values given above.
The
.Fn seed48
function
returns a pointer to an array of 3 shorts which contains the old seed.
This array is statically allocated; thus, its contents are lost after
each new call to
.Fn seed48 .
.Pp
Finally,
.Fn lcong48
allows full control over the multiplicand and addend used in
.Fn drand48 ,
.Fn erand48 ,
.Fn lrand48 ,
.Fn nrand48 ,
.Fn mrand48 ,
and
.Fn jrand48 ,
and the seed used in
.Fn drand48 ,
.Fn lrand48 ,
and
.Fn mrand48 .
An array of 7 shorts is passed as argument; the first three shorts are
used to initialize the seed; the second three are used to initialize the
multiplicand; and the last short is used to initialize the addend.
It is thus not possible to use values greater than 0xffff as the addend.
.Pp
Note that all three methods of seeding the random number generator
always also set the multiplicand and addend for any of the six
generator calls.
.Pp
For a more powerful random number generator, see
.Xr random 3 .
.Sh AUTHORS
.An Martin Birgmeier
.Sh SEE ALSO
.Xr rand 3 ,
.Xr random 3


X11/ChangeLog
2006-09-08  John Harper  <jsh@apple.com>

	* Tagged: LEOPARD_SUBMISSION_0_44

2006-09-06  John Harper  <jsh@apple.com>

	* xc/lib/GL/apple/dri_dispatch.h,
	xc/lib/GL/apple/dri_dispatch.defs: removed glPointParameteri[v]
	aliases

	* xc/programs/Xserver/hw/apple/rootless-window.c,
	xc/programs/Xserver/hw/apple/quartz.c,
	xc/programs/Xserver/hw/apple/applewmstr.h,
	xc/programs/Xserver/hw/apple/applewm.h,
	xc/programs/Xserver/hw/apple/applewm.c,
	xc/programs/Xserver/hw/apple/applewm-impl.h,
	xc/programs/Xserver/hw/apple/applewm-impl.c,
	xc/programs/Xserver/hw/apple/X11Controller.m,
	xc/programs/Xserver/hw/apple/X11Application.m: use shared
	version of AppleWM headers

	* make-host-def, build.sh, Makefile: removed Xplugin/quartz-wm
	references; attempt to support x86_64

2006-07-07  John Harper  <jsh@apple.com>

	* Tagged: LEOPARD_SUBMISSION_0_43

	* xc/programs/Xserver/hw/apple/x-hash.c: fix compiler warning

	* xc/programs/xterm/main.c: fix xterm build failure (4607761)

2006-02-12  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/GL/apple/indirect.c
	(glAquaInitVisualConfigs): add stereo visuals (4377346)

2006-01-16  John Harper  <jsh@apple.com>

	* Tagged: LEOPARD_SUBMISSION_0_42

	* xc/programs/Xserver/hw/apple/bundle-main.c: added /sbin and
	/usr/sbin to the default PATH value (4408133)

	* xc/programs/Xserver/hw/apple/quartz.c (do_exec): use same
	default PATH and MANPATH values as bundle-main.c (4408133)

2005-11-29  John Harper  <jsh@apple.com>

	Fix xscreensaver-demo crash and make its cross-process GL
	surfaces work correctly (4356702):

	* xc/programs/Xserver/hw/apple/dri-surface.c: don't compile the
	code to draw on surfaces (that was always disables but was
	causing a crash - I've fixed the crash, but the code is never
	used so best to just remove it from the binary)

2005-10-25  John Harper  <jsh@apple.com>

	* Tagged: LEOPARD_SUBMISSION_0_41

	* xc/lib/xtrans/Xtranssock.c: fix Leopard build failure
	(4222517)

2005-10-24  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/bundle-main.c: fix default
	MANPATH (4312329)

2005-08-29  John Harper  <jsh@apple.com>

	* xc/extras/freetype2/src/base/ftmac.c: merged endian changes
	from 3477131

2005-06-22  John Harper  <jsh@apple.com>

	* xc/extras/Xpm/lib/scan.c, xc/extras/Xpm/lib/create.c: merged
	fix for CAN-2005-0605 from xfree86 cvs (4031381)

2005-05-24  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/bundle-main.c
	(install_ipaddr_source): changed the way we listen for network
	/ hostname changes, hopefully to be more reliable

	* xc/programs/Xserver/hw/apple/X11Application.m (sendEvent):
	change fix for bring-all-to-front not to require private
	headers - hardcode enum value we need :-(  (4075093)

2005-05-23  John Harper  <jsh@apple.com>

	* Makefile, make-host-def, merge-libs.sh,
	xc/config/cf/Imake.cf, xc/config/cf/darwin.cf,
	xc/config/cf/darwinLib.tmpl, xc/include/Xmd.h,
	xc/lib/Xt/TMaction.c, xc/lib/Xt/Xtos.h,
	xc/lib/freetype2/Imakefile: checked in changes to
	support building fat libraries when ppc64 is in RC_ARCHS
	(3721701)

	* xc/programs/Xserver/hw/apple/X11.pbproj/project.pbxproj:
	bumped short version number to 1.2 so we don't fail to verify

	* xc/programs/Xserver/hw/apple/X11Application.m: handle scroll
	wheel events with delta < 1 correctly (4109092)

2005-05-20  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/rootless.h,
	xc/programs/Xserver/hw/apple/rootless-window.c,
	xc/programs/Xserver/hw/apple/quartz.h,
	xc/programs/Xserver/hw/apple/quartz.c,
	xc/programs/Xserver/hw/apple/X11Application.m: recognize the
	bfCPSOrderAllWindowsForward bit in activation events and
	respond to it appropriately (4075093)

2005-03-09  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_40

	* xc/programs/Xserver/hw/apple/darwin.c
	(DarwinUpdateModifiers): handle AlphaShift (caps lock) as a
	normal modifier, now that we disabled XKB (when merging with
	XFree86 4.4) that's the correct thing to do (4015381)

2005-03-03  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_39

2005-03-01  John Harper  <jsh@apple.com>

	* xc/extras/Xpm/lib/scan.c, xc/extras/Xpm/lib/create.c: more
	Xpm fixes, from
	https://bugs.freedesktop.org/show_bug.cgi?id=1920, attachment
	1909. (4031381)

	* xc/lib/Xpm/Imakefile: doh, forgot Imakefile portion of
	original patch

	* xc/extras/Xpm/lib/scan.c, xc/extras/Xpm/lib/s_popen.c,
	xc/extras/Xpm/lib/parse.c, xc/extras/Xpm/lib/misc.c,
	xc/extras/Xpm/lib/hashtab.c, xc/extras/Xpm/lib/data.c,
	xc/extras/Xpm/lib/create.c, xc/extras/Xpm/lib/XpmI.h,
	xc/extras/Xpm/lib/WrFFrI.c, xc/extras/Xpm/lib/WrFFrBuf.c,
	xc/extras/Xpm/lib/RdFToI.c, xc/extras/Xpm/lib/RdFToBuf.c,
	xc/extras/Xpm/lib/Imakefile, xc/extras/Xpm/lib/CrDatFrI.c,
	xc/extras/Xpm/lib/CrBufFrI.c, xc/extras/Xpm/lib/Attrib.c:
	merged changes from XFree86 HEAD branch (4022757)

2005-02-15  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_38

	* xc/programs/Xserver/hw/apple/quartz-server.c: request that
	deferred updates are disabled. Otherwise we lose 25%
	performance on the "cat file in xterm" test (3985081)

	Fix 3494176 (deadlock while closing OpenGL window):

	* xc/lib/GL/glx/glxext.c: make __glXLock a recursive mutex

	* xc/lib/GL/apple/dri_driver.h, xc/lib/GL/apple/dri_driver.c:
	make the screen mutex a pthread_mutex and make it recursive

2005-01-04  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_37

	* xc/programs/Xserver/hw/apple/X11Controller.m: fixed several
	bugs in the customization dialog for the applications menu
	(3668769)

2004-11-24  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_36_1

	* Makefile: strip-libraries target doesn't strip libXplugin

2004-11-23  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_36

2004-11-16  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/bundle-main.c: added a signal
	handler to catch SIGTERM etc and exit cleanly, killing the X
	server and xinit (3879878)


ksh/ksh/src/lib/libast/uwin/rand48.c
/*******************************************************************
*                                                                  *
*             This software is part of the ast package             *
*                Copyright (c) 1985-2004 AT&T Corp.                *
*        and it may only be used by you under license from         *
*                       AT&T Corp. ("AT&T")                        *
*         A copy of the Source Code Agreement is available         *
*                at the AT&T Internet web site URL                 *
*                                                                  *
*       http://www.research.att.com/sw/license/ast-open.html       *
*                                                                  *
*    If you have copied or used this software without agreeing     *
*        to the terms of the license you are infringing on         *
*           the license and copyright and are violating            *
*               AT&T's intellectual property rights.               *
*                                                                  *
*            Information and Software Systems Research             *
*                        AT&T Labs Research                        *
*                         Florham Park NJ                          *
*                                                                  *
*               Glenn Fowler <gsf@research.att.com>                *
*                David Korn <dgk@research.att.com>                 *
*                 Phong Vo <kpv@research.att.com>                  *
*                                                                  *
*******************************************************************/
#include "FEATURE/uwin"

#if !_UWIN || _lib_srand48

void _STUB_srand48(){}

#else

#define drand48	______drand48
#define erand48	______erand48
#define jrand48	______jrand48
#define lcong48	______lcong48
#define lrand48	______lrand48
#define mrand48	______mrand48
#define nrand48	______nrand48
#define seed48	______seed48
#define srand48	______srand48

#include	<stdlib.h>

#undef	drand48
#undef	erand48
#undef	jrand48
#undef	lcong48
#undef	lrand48
#undef	mrand48
#undef	nrand48
#undef	seed48
#undef	srand48

#if defined(__EXPORT__)
#define extern		__EXPORT__
#endif

#define A	0x5DEECE66D
#define A0	0X5
#define A1	0xDEEC
#define A2	0xE66D
#define C	0xB
#define XINIT	0x330E
#define SCALE	3.55271e-15

static unsigned short oldval[3];
static unsigned short X[3] = { 0, 0, XINIT};
static unsigned short a[3] = { A0, A1, A2};
static unsigned short c = C;

static void multadd(unsigned short x[3], unsigned short a[3], unsigned short c)
{
	register unsigned long r = c;
	unsigned short x2 = x[2];
	unsigned short x1 = x[1];
	r += a[2]*x2;
	x[2] = (unsigned short)r;
	r >>= 16;
	r += a[1]*x2;
	r += a[2]*x1;
	x[1] = (unsigned short)r;
	r >>= 16;
	r += a[2]*x[0];
	r += a[1]*x1;
	r += a[0]*x2;
	x[0] = (unsigned short)r;
}

extern double drand48(void)
{
	double d;
	unsigned long u;
	multadd(X,a,c);
	u = (X[0]<<16) + X[1];
	d = (u*65536.) + X[2];
	return(d*SCALE);
}

extern double erand48(unsigned short xsubi[3])
{
	double d;
	unsigned long u;
	multadd(xsubi,a,c);
	u = (xsubi[0]<<16) + xsubi[1];
	d = (u*65536.) + xsubi[2];
	return(d*SCALE);
}

extern long jrand48(unsigned short xsubi[3])
{
	long u;
	multadd(xsubi,a,c);
	u = (xsubi[0]<<16) | xsubi[1];
	return((long)u);
}

extern void lcong48(unsigned short param[7])
{
	X[0] = param[0];
	X[1] = param[1];
	X[2] = param[2];
	a[0] = param[3];
	a[1] = param[4];
	a[2] = param[5];
	c = param[6];
}

extern long lrand48(void)
{
	long l;
	multadd(X,a,c);
	l = (X[0]<<15)|(X[1]>>1);
	return(l);
}

extern long mrand48(void)
{
	unsigned long u;
	multadd(X,a,c);
	u = (X[0]<<16) | X[1];
	return((long)u);
}

extern long nrand48(unsigned short xsubi[3])
{
	long l;
	multadd(xsubi,a,c);
	l = (xsubi[0]<<15)|(xsubi[1]>>1);
	return(l);
}

extern unsigned short *seed48(unsigned short seed[3])
{
	unsigned short *sp = (unsigned short*)&X;
	a[0] = A0;
	a[1] = A1;
	a[2] = A2;
	c = C;
	oldval[0] = X[2];
	oldval[1] = X[1];
	oldval[2] = X[0];
	X[0] = seed[2];
	X[1] = seed[1];
	X[2] = seed[0];
	return(oldval);
}

extern void srand48(long seedval)
{
	a[0] = A0;
	a[1] = A1;
	a[2] = A2;
	c = C;
	X[0] = (unsigned short)(((unsigned long)seedval) >> 16);
	X[1] = (unsigned short)seedval;
	X[2] = XINIT; 
}

#endif
